legate/core/projection.py

```diff
@@ -12,34 +12,38 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+from __future__ import annotations
 
+from typing import Any, Callable, Optional, Union
 
 from legion_cffi import ffi  # Make sure we only have one ffi instance
 
 
-class ProjExpr(object):
-    def __init__(self, dim=-1, weight=1, offset=0):
+class ProjExpr:
+    def __init__(
+        self, dim: int = -1, weight: int = 1, offset: int = 0
+    ) -> None:
         self._dim = dim
         self._weight = weight
         self._offset = offset
-        self._repr = None
+        self._repr: Union[str, None] = None
```

Comment:
What's a typical way to express a nullable type in the Python typing parlance? Coming from a PL background, I'd find `Optional[str]` more comfortable and `Union` reserved for a sum type, though I guess `Union[str, None]` is equivalent. I'm just curious to know if there's any reason to prefer `Union` over `Optional` for cases like this.

Reply:
`Optional[str]` would certainly work here. I personally prefer to reserve `Optional` for parameters with default None values, i.e. where it is optional for a caller to pass a value or not. To me, a value like this is not optional in that sense, it simply a union type that can hold either `str` or `None`. This is not a super strong opinion, though, so I am happy to change all `Union[..., None]` to `Optional[...]` if that's what others prefer.

Reply:
I guess I'm fine with `Union`, as long as the same convention is used throughout the code, so no need for changing the code for this.

Reply:
:+1:  In the future I would actually advocate for `Foo | None` everywhere (less to type, less to import)